I tried testing these changes to run the EnergyReco/Predict toolchain. I do:
`make -j4 EnergyReco/Predict`
the TOOLS needed and OBJECTS are printed out properly for this specific Toolchain only but then I get this message:
`make: *** No rule to make target 'EnergyReco/Predict'.  Stop.`
I tried running another toolchain (i.e BeamFetcherV2) with
`make -j4 BeamFetcherV2`
 to make sure this is not specific to this toolchain and I didn't get this message. However, I got this message:

> TOOLCHAIN: ./configfiles/BeamFetcherV2
> TOOLS needed: LoadGeometry LoadRawData TriggerDataDecoder BeamFetcherV2
> OBJECTS: UserTools/LoadGeometry/LoadGeometry.o UserTools/LoadRawData/LoadRawData.o UserTools/TriggerDataDecoder/TriggerDataDecoder.o UserTools/BeamFetcherV2/BeamFetcherV2.o UserTools/BeamFetcherV2/IFBeamDBInterfaceV2.o
> if [[ `head -n1 UserTools/MyFactory/Unity.h | sed s://::` != "./configfiles/BeamFetcherV2" ]]; then rm -rf UserTools/MyFactory/Unity.h  UserTools/MyFactory/MyFactory.cpp UserTools/MyFactory/MyFactory.o; fi
> head: cannot open 'UserTools/MyFactory/Unity.h' for reading: No such file or directory
> making NEW
> making ToolChain ./configfiles/BeamFetcherV2

and it started compiling the tools but it suddenly stopped. The message below hints that there is a conflict introduced with how the name of the toolchain is retrieved.

> make[1]: *** No rule to make target 'configfiles/./configfiles/BeamFetcherV2/ToolsConfig', needed by 'UserTools/MyFactory/Unity.h'.  Stop.
> make[1]: Leaving directory '/ToolAnalysisLink_test'
> make: *** [Makefile:244: BeamFetcherV2] Error 2

I think this can be fixed if we remove `configfiles` before $(TOOLCHAIN) wherever it appears in the Makefile (if nothing else is broken). However, I still  can't figure out the first issue with EnergyReco/Predict toolchain. Could you see if you can reproduce these? Should we check these before merging?

--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,5 @@
 ToolDAQPath=${PWD}/ToolDAQ
 
-ifeq ($(TOOLCHAIN),)
 TOOLCHAIN:=$(shell find ./configfiles -type d -path ./configfiles/$(MAKECMDGOALS) -exec echo {} \; 2>/dev/null)
 TOOLCHAINNAME:=$(subst ./configfiles/,,$(TOOLCHAIN))
 $(info TOOLCHAINNAME: $(TOOLCHAINNAME))
@@ -16,7 +15,6 @@ NPROCS:=$(shell nproc --all)
 GIT_VERSION := "$(shell git describe --dirty --always)"
 # forward these to variables to internal $(MAKE) calls
 export
-endif
 
 ifeq ($(TOOLCHAIN),)
 TOOLCHAIN:=Dummy

--- a/Makefile
+++ b/Makefile
@@ -2,9 +2,13 @@ ToolDAQPath=${PWD}/ToolDAQ
 
 ifeq ($(TOOLCHAIN),)
 TOOLCHAIN:=$(shell find ./configfiles -type d -path ./configfiles/$(MAKECMDGOALS) -exec echo {} \; 2>/dev/null)
+TOOLCHAINNAME:=$(subst ./configfiles/,,$(TOOLCHAIN))
+$(info TOOLCHAINNAME: $(TOOLCHAINNAME))
+ifneq ($(MAKECMDGOALS),clean)
 TOOLS:=$(shell ./gettools.sh $(TOOLCHAIN))
 #OBJECTS:=$(foreach tool,$(TOOLS),UserTools/$(tool)/$(tool).o)
 OBJECTS:=$(shell ./getobjects.sh $(TOOLS))
+endif
 $(info TOOLCHAIN: $(TOOLCHAIN))
 $(info TOOLS needed: $(TOOLS))
 $(info OBJECTS: $(OBJECTS))
@@ -207,7 +211,7 @@ DataModel/%.o: DataModel/%.cpp lib/libLogging.so lib/libStore.so include/Tool.h
 	@echo -e "\n*************** Making " $@ "****************"
 	-$(CCC) -c -o $@ $< -I include -L lib -lStore -lLogging  $(DataModelInclude) $(DataModelLib) $(ZMQLib) $(ZMQInclude) $(BoostLib) $(BoostInclude)
 
-UserTools/MyFactory/Unity.h: configfiles/$(TOOLCHAIN)/ToolsConfig
+UserTools/MyFactory/Unity.h: $(TOOLCHAIN)/ToolsConfig
 	@echo "Generating $@ for ToolChain $(TOOLCHAIN)"
 	@mkdir -p UserTools/MyFactory
 	@echo "//$(TOOLCHAIN)" > $@
@@ -218,7 +222,7 @@ UserTools/MyFactory/Unity.h: configfiles/$(TOOLCHAIN)/ToolsConfig
 
 
 
-UserTools/MyFactory/MyFactory.cpp: UserTools/MyFactory/Unity.h configfiles/$(TOOLCHAIN)/ToolsConfig
+UserTools/MyFactory/MyFactory.cpp: UserTools/MyFactory/Unity.h $(TOOLCHAIN)/ToolsConfig
 	@echo "Generating $@ for ToolChain $(TOOLCHAIN)"
 	@cp UserTools/Factory/Factory.h UserTools/MyFactory/Factory.h
 	@echo '#include "Factory.h"' > $@
@@ -230,8 +234,9 @@ UserTools/MyFactory/MyFactory.cpp: UserTools/MyFactory/Unity.h configfiles/$(TOO
 	echo 'return ret;' >> $@;\
 	echo '}' >> $@
 
-%:: configfiles/%/ToolsConfig
-	if [[ `head -n1 UserTools/MyFactory/Unity.h | sed s://::` != "$(TOOLCHAIN)" ]]; then rm -rf UserTools/MyFactory/Unity.h  UserTools/MyFactory/MyFactory.cpp UserTools/MyFactory/MyFactory.o; fi
+$(TOOLCHAINNAME):: $(TOOLCHAIN)/ToolsConfig
+	@echo "Target $@"
+	@if [[ `head -n1 UserTools/MyFactory/Unity.h 2>/dev/null | sed s://::` != "$(TOOLCHAIN)" ]]; then rm -rf UserTools/MyFactory/Unity.h  UserTools/MyFactory/MyFactory.cpp UserTools/MyFactory/MyFactory.o; fi
 	@echo "making NEW"
 	@if [ ! -z "$(TOOLCHAIN)" ]; then echo "making ToolChain $(TOOLCHAIN)";fi
 	@# check tools file exists

--- a/Makefile
+++ b/Makefile
@@ -17,13 +17,14 @@ GIT_VERSION := "$(shell git describe --dirty --always)"
 export
 
 ifeq ($(TOOLCHAIN),)
-TOOLCHAIN:=Dummy
+TOOLCHAIN:=./configfiles/Dummy
+TOOLCHAINNAME:=Dummy
 endif
 
 CPPFLAGS= -fmax-errors=1 -DVERSION=\"$(GIT_VERSION)\" -Wno-reorder -Wno-sign-compare -Wno-unused-variable -Wno-unused-but-set-variable -Werror=return-type -Wl,--no-as-needed
 
-CC=g++ -std=c++1y -O3 -fPIC -shared $(CPPFLAGS)
-CCC= g++ -std=c++1y -O3 -fPIC  $(CPPFLAGS)
+CC=g++ -std=c++1y -g -O3 -fPIC -shared $(CPPFLAGS)
+CCC= g++ -std=c++1y -g -O3 -fPIC  $(CPPFLAGS)
 
 
 ZMQLib= -L $(ToolDAQPath)/zeromq-4.0.7/lib -lzmq

--- a/gettools.sh
+++ b/gettools.sh
@@ -19,7 +19,9 @@ if [ $# -eq 0 ] || [ ! -f "configfiles/$TOOLCHAIN/ToolsConfig" ]; then
 		TOOLS+=( "${TOOLNAME}" )
 	done
 	
-	echo "${TOOLS[@]}"
+	#echo "${TOOLS[@]}"
+	UNIQUE_TOOLS=( $(for ATOOL in "${TOOLS[@]}"; do echo "${ATOOL}"; done | sort -u) )
+	echo "${UNIQUE_TOOLS[@]}"
 	#ls UserTools/*/*.cpp | sed 's/.cpp$//' | xargs -n1 basename | grep -v -e Factory -e template -e InactiveTools
 else
 	#echo "getting tools needed for toolchain $TOOLCHAIN" >&2
@@ -33,5 +35,7 @@ else
 		TOOLS+=( "${TOOL}" )
 		#echo "adding tool ${TOOL}"  >&2
 	done < <(cat "./configfiles/$TOOLCHAIN/ToolsConfig")
-	echo "${TOOLS[@]}"
+	#echo "${TOOLS[@]}"
+	UNIQUE_TOOLS=( $(for ATOOL in "${TOOLS[@]}"; do echo "${ATOOL}"; done | sort -u) )
+	echo "${UNIQUE_TOOLS[@]}"
 fi

